# TASK 1: DATA EXTRACTION AND INITIAL ANALYSIS

In [1]:
import pandas as pd

In [2]:
df = pd.read_excel('financials_2022_2024.xlsx')
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0  Microsoft         2024         245122       88136        512163   
1  Microsoft         2023         211915       72361        411976   
2  Microsoft         2022         198270       72738        364840   
3      Tesla         2024          97690        7153        122070   
4      Tesla         2023          96773       14974        106618   
5      Tesla         2022          81462       12587         82338   
6      Apple         2024         394328       93736        364980   
7      Apple         2023         365817       96995        352583   
8      Apple         2022         274515       99803        352755   

   Total Liabilities  Cash Flow from Operating Activities  
0             243686                               118548  
1             205753                                87582  
2             198298                                89035  
3              48390                                14923  
4              43009                                13256  
5              36440                                14724  
6             308030                               118254  
7             290437                               110543  
8             302083                               122151

## Create copy

In [3]:
df2 = df.copy()

In [4]:
df2

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0  Microsoft         2024         245122       88136        512163   
1  Microsoft         2023         211915       72361        411976   
2  Microsoft         2022         198270       72738        364840   
3      Tesla         2024          97690        7153        122070   
4      Tesla         2023          96773       14974        106618   
5      Tesla         2022          81462       12587         82338   
6      Apple         2024         394328       93736        364980   
7      Apple         2023         365817       96995        352583   
8      Apple         2022         274515       99803        352755   

   Total Liabilities  Cash Flow from Operating Activities  
0             243686                               118548  
1             205753                                87582  
2             198298                                89035  
3              48390                                14923  
4              43009                                13256  
5              36440                                14724  
6             308030                               118254  
7             290437                               110543  
8             302083                               122151

In [6]:
# Set company as the index
df2.set_index('Company')

Fiscal Year  Total Revenue  Net Income  Total Assets  \
Company                                                           
Microsoft         2024         245122       88136        512163   
Microsoft         2023         211915       72361        411976   
Microsoft         2022         198270       72738        364840   
Tesla             2024          97690        7153        122070   
Tesla             2023          96773       14974        106618   
Tesla             2022          81462       12587         82338   
Apple             2024         394328       93736        364980   
Apple             2023         365817       96995        352583   
Apple             2022         274515       99803        352755   

           Total Liabilities  Cash Flow from Operating Activities  
Company                                                            
Microsoft             243686                               118548  
Microsoft             205753                                87582  
Microsoft             198298                                89035  
Tesla                  48390                                14923  
Tesla                  43009                                13256  
Tesla                  36440                                14724  
Apple                 308030                               118254  
Apple                 290437                               110543  
Apple                 302083                               122151

## Sort values to show from 2022 and Average and totals by company in the past 3 fiscal years

In [7]:
df2 = df2.sort_values(['Company', 'Fiscal Year']).reset_index(drop=True)

In [8]:
df2

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2022         274515       99803        352755   
1      Apple         2023         365817       96995        352583   
2      Apple         2024         394328       93736        364980   
3  Microsoft         2022         198270       72738        364840   
4  Microsoft         2023         211915       72361        411976   
5  Microsoft         2024         245122       88136        512163   
6      Tesla         2022          81462       12587         82338   
7      Tesla         2023          96773       14974        106618   
8      Tesla         2024          97690        7153        122070   

   Total Liabilities  Cash Flow from Operating Activities  
0             302083                               122151  
1             290437                               110543  
2             308030                               118254  
3             198298                                89035  
4             205753                                87582  
5             243686                               118548  
6              36440                                14724  
7              43009                                13256  
8              48390                                14923

## Group the total revenues and net incomes for better analysis

In [9]:
# Average and totals by company in the 3 years
summary = df2.groupby('Company').agg({
    'Total Revenue': ['mean', 'max', 'min'],
    'Net Income': ['mean', 'max', 'min'],
    'Total Assets': 'mean',
    'Total Liabilities': 'mean',
    'Cash Flow from Operating Activities': 'mean'
})
print(f"\n--- IN MILLIONS ---")
summary.round(2)


--- IN MILLIONS ---


Total Revenue                 Net Income               Total Assets  \
                   mean     max     min       mean    max    min         mean   
Company                                                                         
Apple         344886.67  394328  274515   96844.67  99803  93736    356772.67   
Microsoft     218435.67  245122  198270   77745.00  88136  72361    429659.67   
Tesla          91975.00   97690   81462   11571.33  14974   7153    103675.33   

          Total Liabilities Cash Flow from Operating Activities  
                       mean                                mean  
Company                                                          
Apple             300183.33                           116982.67  
Microsoft         215912.33                            98388.33  
Tesla              42613.00                            14301.00

In [11]:
# Sort 
df2.sort_values(['Company', 'Fiscal Year'])

# Group by company to see the total revenue and net incomes 
for company, group in df2.groupby('Company'):
    print(f"\n--- {company} IN MILLIONS---")
    print(group[['Fiscal Year', 'Total Revenue', 'Net Income']])


--- Apple IN MILLIONS---
   Fiscal Year  Total Revenue  Net Income
0         2022         274515       99803
1         2023         365817       96995
2         2024         394328       93736

--- Microsoft IN MILLIONS---
   Fiscal Year  Total Revenue  Net Income
3         2022         198270       72738
4         2023         211915       72361
5         2024         245122       88136

--- Tesla IN MILLIONS---
   Fiscal Year  Total Revenue  Net Income
6         2022          81462       12587
7         2023          96773       14974
8         2024          97690        7153


## To analyze the revenue growth (%) and net income growth (%)

In [14]:
df2['Revenue Growth (%)'] = (
    df2.groupby('Company')['Total Revenue']
      .pct_change()
      .fillna(0)
      * 100
)
df2['Net Income Growth (%)'] = (
    df2.groupby('Company')['Net Income']
      .pct_change()
      .fillna(0)
      * 100
)
df_growth = df2[['Company','Fiscal Year','Total Revenue','Revenue Growth (%)','Net Income','Net Income Growth (%)']]
df_growth.round(2)

Company  Fiscal Year  Total Revenue  Revenue Growth (%)  Net Income  \
0      Apple         2022         274515                0.00       99803   
1      Apple         2023         365817               33.26       96995   
2      Apple         2024         394328                7.79       93736   
3  Microsoft         2022         198270                0.00       72738   
4  Microsoft         2023         211915                6.88       72361   
5  Microsoft         2024         245122               15.67       88136   
6      Tesla         2022          81462                0.00       12587   
7      Tesla         2023          96773               18.80       14974   
8      Tesla         2024          97690                0.95        7153   

   Net Income Growth (%)  
0                   0.00  
1                  -2.81  
2                  -3.36  
3                   0.00  
4                  -0.52  
5                  21.80  
6                   0.00  
7                  18.96  
8                 -52.23

# CONCLUSION

Apple:
Revenue grew steadily from $274.5B to $394.3B (+41%), though growth slowed in 2024. Net income declined slightly each year (−2.8% in 2023, −3.4% in 2024), suggesting tighter margins despite strong sales.

Microsoft:
Revenue rose consistently from $198.3B to $245.1B (+24%). Net income dipped slightly in 2023 but rebounded strongly in 2024 (+21.8%), showing renewed profitability and efficient cost management.

Tesla:
Revenue increased from $81.5B to $97.7B, but growth nearly stalled in 2024 (+0.9%). Net income fell sharply (−52% in 2024), indicating major margin pressure likely due to higher costs or price reductions.